### ISSUE TITLE
`operator.getitem` doesn't like being passed a key of type `SupportsIndex`

### ISSUE THREAD

**Issue by posita** [python/typeshed#10519]
``` py
# test_case.py
import operator
from typing import Sequence, SupportsIndex, Union

GetItemT = Union[SupportsIndex, slice]

class Spam:
    pass

def okay1(seq: Sequence[Spam], key: int) -> Spam:
    return operator.__getitem__(seq, key)  # fine

def okay2(seq: Sequence[Spam], key: slice) -> Sequence[Spam]:
    return operator.__getitem__(seq, key)  # still fine

def error1(seq: Sequence[Spam], key: SupportsIndex) -> Spam:
    return operator.__getitem__(seq, key)  # <- oh no!

def error2(seq: Sequence[Spam], key: GetItemT) -> Union[Spam, Sequence[Spam]]:
    return operator.__getitem__(seq, key)  # <- unsurprising given prior error (but the error message seems misleading)
```

```
% mypy --version
mypy 1.4.1 (compiled: yes)
% mypy --config=/dev/null test_case.py
/dev/null: No [mypy] section in config file
test_case.py:17: error: Cannot infer function type argument  [misc]
test_case.py:20: error: Argument 2 has incompatible type "SupportsIndex | slice"; expected "slice"  [arg-type]
```

:point_up: Same result with both Python 3.9 and 3.10 and with `... | ...` notation (instead of `Union[..., ...]`) in 3.10. `s/int/SupportsIndex/g` clears up the error. [This](https://github.com/python/typeshed/blob/0d8a6bc2004b7b6d8335e35b017438125c305264/stdlib/_operator.pyi#L80C1-L84) is a bit hard to wrap my head around, so I'm not quite clear on the source of the error.

**Comment by sobolevn** [comment]
I think that you can experiment with adding 

```python
@overload
def getitem(__a: Sequence[_T], __b: SupportsIndex) -> _T: ...
```

overload to the definition of `getitem`.

Adding tests to `test_cases/` here would be required as well :)

**Comment by JelleZijlstra** [comment]
This isn't related to the `operator` module, it's because `Sequence.__getitem__` only accepts `int`, not `SupportsIndex`.

**Comment by AlexWaygood** [comment]
I agree with Jelle on this.

Allowing `operator.getitem` to be used with `Sequence` and `SupportsIndex` in this way would be unsafe unless we also changed the signature of `typing.Sequence.__getitem__`. Otherwise mypy would let you do this:

```py
import operator
from collections.abc import Sequence
from typing import SupportsIndex, overload

class MySequence(Sequence[bool]):
    @overload
    def __getitem__(self, x: int) -> bool: ...
    @overload
    def __getitem__(self, x: slice) -> Sequence[bool]: ...
    def __getitem__(self, x: int | slice) -> bool | Sequence[bool]:
        if isinstance(x, int):
            return x.bit_length() > 10
        return [False for _ in range(x.start, x.stop)]
    def __len__(self) -> int:
        return 42


class HasIndexMethod:
    def __index__(self) -> int:
        return 42


def test(x: MySequence, y: SupportsIndex) -> None:
    operator.getitem(x, y)


test(MySequence(), HasIndexMethod())  # boom!
```

Which fails at runtime with:

```pytb
Traceback (most recent call last):
  File "/Users/alexw/dev/foo.py", line 28, in <module>
    test(MySequence(), HasIndexMethod())  # boom!
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/alexw/dev/foo.py", line 25, in test
    operator.getitem(x, y)
  File "/Users/alexw/dev/foo.py", line 14, in __getitem__
    return [False for _ in range(x.start, x.stop)]
                                 ^^^^^^^
AttributeError: 'HasIndexMethod' object has no attribute 'start'
```

So the obvious next question is, "Okay, should we change `Sequence.__getitem__` so that it accepts `SupportsIndex` instead of `int` for the first overload here?

https://github.com/python/typeshed/blob/704fa5d1decd72857d9b22378abfcf4bfbc00bdd/stdlib/typing.pyi#L582

But the answer is "no", we shouldn't do that either. Although the vast majority of sequences provided by the stdlib do accept any object with an `__index__` method for their `__getitem__` implementations, it would be unduly restrictive for us to define the interface in that way. Arbitrary subclasses of `collections.abc.Sequence` shouldn't have to be able to take arbitrary objects with `__index__` methods to be considered valid implementations of the interface; as long as they can handle `int` (which is the argument passed 99% of the time), they should be fine.

In summary: the existing types seem correct to me; I don't think we should change anything here.

**Comment by posita** [comment]
Fair, but I struggle to see the purpose of the `SupportsIndex` type then? From [`object.__index__(self)[¶](https://docs.python.org/3/reference/datamodel.html#object.__index__)`](https://docs.python.org/3/reference/datamodel.html#object.__index__) (**_emphasis_** added):

> Called to implement [`operator.index()`](https://docs.python.org/3/library/operator.html#operator.index), and whenever **_Python_** needs to losslessly convert the numeric object to an integer object (**_such as in slicing_**, or in the built-in [`bin()`](https://docs.python.org/3/library/functions.html#bin), [`hex()`](https://docs.python.org/3/library/functions.html#hex) and [`oct()`](https://docs.python.org/3/library/functions.html#oct) functions). Presence of this method indicates that the numeric object is an integer type. Must return an integer.

If you can't pass `SupportsIndex` to `__getitem__` (because `__getitem__` didn't contemplate receiving a `SupportsIndex`, despite the language above), I think that should probably up updated to read, "... whenever ~Python~ _a caller_ needs to _manually_ losslessly convert the numeric object to an integer object (such as ~in slicing~ _when creating a slice_,_ ...) ... ."

The current docs make it sound like it's a holistically (or at least aspirationally) supported interface throughout the standard library.* If that's valid, then I'd suggest there's a bug in your example, which probably should have read:

```py
class MySequence(Sequence[bool]):
    # ...
    def __getitem__(self, x: int | SupportsIndex | slice) -> bool | Sequence[bool]:
        if isinstance(x, int):
            return x.bit_length() > 10
        elif isinstance(x, SupportsIndex):
            return operator.__index__(x).bit_length() > 10
        return [False for _ in range(x.start, x.stop)]
```

Technically, that first `if` branch is redundant since `int` also satisfies `SupportsIndex` (thereby lending credibility to its apparently now-defunct legitimacy). Yes, I understand that puts the onus on implementers of `__getitem__` to support `SupportsIndex`, but if that's not legit, then `SupportsIndex` seems like an abandoned relic of yet another ready-fire-aim typing decision. It should either be supported or explicitly abandoned. It should not linger as a sharp edge on which careful readers cut themselves after having misplaced their trust in canonical documentation.

**Comment by JelleZijlstra** [comment]
You can pass a SupportsIndex to the `__getitem__` method of many concrete sequences (e.g., list, str). You just can't pass it to an *arbitrary* sequence.
